doc/kruskal.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "KRUSKAL" "1" "September 2017" "www.complex-networks.net" "www.complex-networks.net"
.
.SH "NAME"
\fBkruskal\fR \- Find the minimum/maximum spanning tree of a graph
.
.SH "SYNOPSIS"
\fBkruskal\fR \fIgraph_in\fR [MAX]
.
.SH "DESCRIPTION"
\fBkruskal\fR computes the minimum (or maximum) spanning tree of \fIgraph_in\fR, using the Kruskal\'s algorithm\. If \fIgrahp_in\fR is unweighted, \fBkruskal\fR computes one of the spanning trees of the graph\. The program prints on output the (weighted) edge list of the spanning tree\.
.
.SH "PARAMETERS"
.
.TP
\fIgraph_in\fR
undirected input graph (edge list)\. It must be an existing file\.
.
.TP
\fBMAX\fR
If the second (optional) parameter is equal to \fBMAX\fR, compute the maximum spanning tree\. Otherwise, compute the minimum spanning tree\.
.
.SH "OUTPUT"
The program prints on STDOUT the edge list of the minimum (maximum) spannig tree of \fIgraph_in\fR, in the format:
.
.IP "" 4
.
.nf

    i_1 j_1 w_ij_1
    i_2 j_2 w_ij_2
    \.\.\.\.
.
.fi
.
.IP "" 0
.
.SH "EXAMPLES"
To find the minimum spanning tree of the graph \fBstocks_62_weight\.net\fR (the network of stocks in the New York Exchange market) we use the command:
.
.IP "" 4
.
.nf

    $ kruskal stocks_62_weight\.net
    52 53 0\.72577357
    43 53 0\.72838212
    2 53 0\.72907212
    \.\.\.
    36 53 0\.7973488
    53 58 0\.79931683
    26 27 0\.8029602
    $
.
.fi
.
.IP "" 0
.
.P
which prints on output the edge list of the minimum spanning tree\. However, since the weight of each edge in that graph indicates the similarity in the behaviour of two stocks, the maximum spanning tree contains information about the backbone of similarities among stocks\. To obtain the maximum spannin tree, we just specify \fBMAX\fR as second parameter:
.
.IP "" 4
.
.nf

    $ kruskal stocks_62_weight\.net MAX
    56 58 1\.523483
    2 52 1\.3826744
    32 51 1\.3812241
    \.\.\.
    33 55 0\.86880272
    7 28 0\.8631584
    1 53 0\.81876166
    $
.
.fi
.
.IP "" 0
.
.SH "SEE ALSO"
clust_w(1), dijkstra(1), largest_component(1)
.
.SH "REFERENCES"
.
.IP "\(bu" 4
J\. B\. Kruskal\. "On the shortest spanning subtree of a graph and the traveling sales\-man problem"\. P\. Am\. Math\. Soc\. 7 (1956), 48\-48\.
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Appendix 20, Cambridge University Press (2017)
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Chapter 10, Cambridge University Press (2017)
.
.IP "" 0
.
.SH "AUTHORS"
(c) Vincenzo \'KatolaZ\' Nicosia 2009\-2017 \fB<v\.nicosia@qmul\.ac\.uk>\fR\.


debian/changelog
netbunch (1.0-1) unstable; urgency=medium

  * Initial release 

 -- Vincenzo (KatolaZ) Nicosia <katolaz@freaknet.org>  Thu, 28 Sep 2017 06:46:47 +0100


src/kruskal/kruskal.c
/**
 *  This program is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 *  (c) Vincenzo Nicosia 2009-2017 -- <v.nicosia@qmul.ac.uk>
 * 
 *  This file is part of NetBunch, a package for complex network
 *  analysis and modelling. For more information please visit:
 *
 *             http://www.complex-networks.net/
 *
 *  If you use this software, please add a reference to 
 *
 *               V. Latora, V. Nicosia, G. Russo             
 *   "Complex Networks: Principles, Methods and Applications"
 *              Cambridge University Press (2017) 
 *                ISBN: 9781107103184
 *
 ***********************************************************************
 *
 *  This program finds the minimum/maximum spanning tree of a graph
 *  given as input, using the Kruskal's algorithm 
 *
 *
 *  References:
 *  
 * [1] J. B. Kruskal. "On the shortest spanning subtree of a graph and
 *     the traveling sales-man problem". P. Am. Math. Soc. 7 (1956),
 *     48-48.  
 *
 */



#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "dset.h"
#include "gen_heap.h"
#include "utils.h"
#include "edge_w_funs.h"


void usage(int argc, char *argv[]){

  printf("********************************************************************\n"
         "**                                                                **\n"
         "**                        -*-   kruskal  -*-                      **\n"
         "**                                                                **\n"
         "**   Find the minimum/maximum spanning tree of a weighted graph   **\n"
         "**   'graph_in' provided as input.                                **\n"
         "**                                                                **\n"
         "**   The input file 'graph_in' is a weighted edge-list:           **\n"
         "**                                                                **\n"
         "**                        I_1 J_1 W_1                             **\n"
         "**                        I_2 J_2 W_2                             **\n"
         "**                        I_3 J_3 W_3                             **\n"
         "**                        ... ...                                 **\n"
         "**                        I_K J_K W_K                             **\n"
         "**                                                                **\n"
         "**    The program computes by default the minimum spanning tree   **\n"
         "**    of 'graph_in', unless the second parameter 'MAX' is         **\n"
         "**    specified.                                                  **\n"
         "**                                                                **\n"
         "**    The program prints on STDOUT the weighted edge-list of the  **\n"
         "**    minimum/maximum spanning tree of 'graph_in'.                **\n"
         "**                                                                **\n"
         "**    If 'graph_in' is an unweighted graph, the program prints    **\n"
         "**    on output one of the spanning trees of the graph.           **\n"
         "**                                                                **\n"
         "********************************************************************\n"
         " This is Free Software - You can use and distribute it under \n"
         " the terms of the GNU General Public License, version 3 or later\n\n"
         " Please visit http://www.complex-networks.net for more information\n\n"
         " (c) Vincenzo Nicosia 2009-2017 (v.nicosia@qmul.ac.uk)\n"
         "********************************************************************\n\n"
         );
  printf("Usage: %s <graph_in> [MAX]\n", argv[0]);
}


void kruskal(gen_heap_t *h, unsigned int N){
  
  dset_t *nodes, set1, set2;
  edge_w_t *elem;
  unsigned int i;
  
  
  nodes = malloc(N * sizeof(dset_t));
  
  for(i=0; i<N; i++){
    nodes[i] = NULL;
    dset_makeset(nodes + i);
  }
  
  while(! gen_heap_delete(h, (void**)(&elem))){
    /* get the next edge */
    set1 = dset_find(nodes[elem->i]);
    set2 = dset_find(nodes[elem->j]);
    /* if i and j do not belong to the same disjoint set...*/
    if (set1 != set2){
      /* ... the edge (i,j) belongs to the spanning tree, */
      /* so we print it...*/
      print_elem_edge_w(elem);
      /* ...then merge the two sets... */
      dset_union(nodes[elem->i], nodes[elem->j]);
    }
    free(elem);
  }
  /* We now destroy all the disjoint sets... */
  for (i=0;i<N;i++){
    dset_destroy(nodes[i]);
  }
  free(nodes);
  
}

void load_edges_into_heap(FILE *filein, gen_heap_t *h){
  
  char buff[256];
  char *ptr;
  edge_w_t *elem;
  
  
  while(fgets(buff, 256, filein)){
    if (buff[0] == '#')
      continue;
    elem = malloc(sizeof(edge_w_t));
    ptr = strtok(buff, " "); /* read the first node */
    VALID_PTR_OR_EXIT(ptr, 7);
    elem->i = atoi(ptr);
    ptr = strtok(NULL, " "); /* read the second node */
    VALID_PTR_OR_EXIT(ptr, 7);
    elem->j = atoi(ptr);
    ptr = strtok(NULL, " "); /* read the weight */
    if (!ptr)
      /* if no weight is specified, assume it is set to 1.0 */
      elem->w = 1.0;
    else
      elem->w = atof(ptr);
    /* put the edge in the heap */
    gen_heap_insert(h, elem);
  }
  
}

int count_num_lines(FILE *filein){
  
  int i, ch ;

  i = 0;
  
  while ((ch = fgetc(filein)) != EOF){
      if (ch == '\n')
        i ++;
  }
  rewind(filein);
  return i;
}


int main(int argc, char *argv[]){
  
  gen_heap_t *h;
  unsigned int N;
  FILE *filein;
  char htype;
  gen_heap_func_t  *funs;
  
  if(argc < 2){
    usage(argc, argv);
    exit(1);
  }

  htype = MIN_HEAP;
  if (argc > 2 && !my_strcasecmp(argv[2], "MAX")){
    htype = MAX_HEAP;
  }

  /* Initialisation of functions for gen_heap*/
  funs = malloc(sizeof(gen_heap_func_t));
  (*funs).compare = compare_edge_w;
  (*funs).alloc_vector = alloc_vector_edge_w;
  (*funs).dealloc_vector = dealloc_vector_edge_w;
  (*funs).dealloc_elem = dealloc_elem_edge_w;
  (*funs).print_elem = print_elem_edge_w;


  
  filein = openfile_or_exit(argv[1], "r", 2);
  
  N = count_num_lines(filein);

  h = gen_heap_init(N, htype, funs);
  
  load_edges_into_heap(filein, h);

  fclose(filein);
  
  kruskal(h, N);
  gen_heap_destroy(h);
  free(funs);
}
